pkg/controller/name.go

```diff
@@ -0,0 +1,43 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+var nameLock sync.Mutex
+var usedNames sets.Set[string]
```

Comment:
Hi @alvaroaleman @sbueringer 
Holding controller name with a global variable seems not a good idea, because the variable will be shared between tests in a test suite or in a test package. 

To be more specific, see following the test sample below:
```golang
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			mgr, err := controllerruntime.NewManager(&rest.Config{}, controllerruntime.Options{})

			controller := &FooController{}
			err = controller.SetupWithManager(mgr)
			assert.NoError(t, err) // raise error: Controller names must be unique to avoid multiple controllers reporting to the same metric

			// ... others
		})
	}
```
Even though each test creates a brand new controller manager, but the `usedNamed` is shared unexpected.

Reply:
My two cents:
The used controller names should be stored in a specific manager instance.

Reply:
@RainbowMango my tests have the following: `ctrl.NewManager(cfg, ctrl.Options{Scheme: scheme.Scheme, Metrics: metricsserver.Options{BindAddress: "0"}, Controller: config.Controller{SkipNameValidation: &yes}})`

Reply:
Yeah, thanks. @mythi 
It's not a blocker for me.
I noticed the option as well, but I think it's more like a workaround.